PurchaseNotice: Add alternative expiry text for temporary site purchases

mattgawarecki opened the pull request:
This pull request adds alternative text to the purchase expiry notice for siteless purchases. The new text still says the purchase will expire and be removed, but does not mention "from this site," since it's not associated with a site.

Resolves `1202619025189113-as-1203845960301499`.

## Proposed Changes

* Add two new forms of expiry text in `getExpiryText`, which are only displayed when the current purchase evaluates to true when passed to `isTemporarySitePurchase`.

## Testing Instructions

* In your Calypso Blue test environment, visit `/me/purchases` and select a purchase that is not associated with any site (e.g., an unassigned Jetpack Boost license).
* Verify you see an expiration notice at the top of the purchase details as shown in the screenshots below. Notably, the text should not mention "this site."

### Reference screenshots

#### Before

<img width="1053" alt="image" src="https://github.com/Automattic/wp-calypso/assets/670067/424a2b43-1b3b-446e-9d73-449a3030f15f">

#### After

<img width="1052" alt="image" src="https://github.com/Automattic/wp-calypso/assets/670067/d8f332a5-ed91-4d7d-b4f2-4662b5aca78c">

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203845960301499

Commit:
PurchaseNotice: Add alternative expiry text for temporary site purchases

sirbrillig commented on client/me/purchases/manage-purchase/notices.tsx:
```diff
@@ -103,6 +103,15 @@ class PurchaseNotice extends Component<
 			const daysToExpiry = moment( expiry.diff( moment() ) ).format( 'D' );
 
+			if ( isTemporarySitePurchase( purchase ) ) {
+				return translate( '%(purchaseName)s will expire and be removed %(expiry)s days. ', {
```
There's a space at the end of this string that probably shouldn't be there. (If we need the space for alignment purposes, we could add it outside the translated string.)

Also, I think the wording here is a little off. "Product will expire and be removed 5 days" maybe should be "Product will expire and be removed **in** 5 days"?

sirbrillig replied:
Unless... does `moment` somehow handle adding the "in" itself?

mattgawarecki replied:
I'm not certain myself, but testing locally, it seems it... _does_ add the 'in' for cases where the expiry is framed in terms of days? This doesn't make any sense to me, TBH, but manually changing the expiry date for my purchase still gives me the correct message. Below is a screenshot after I changed it to a date 14 days in the future. But since this is such unusual behavior (IMO), I encourage anybody testing to double-check me.

<img width="1059" alt="image" src="https://github.com/Automattic/wp-calypso/assets/670067/f5068bf3-cd13-45cb-92cb-0ec74a59e6b8">

sirbrillig replied:
That's... very strange! 🙃  I tried this in a playground and this code definitely does not add 'in':

```js
const expiry = moment( '2023-10-05 00:00Z' );
console.log( moment( expiry.diff( moment() ) ).format( 'D' ) );
```

In fact, it doesn't return a number of days at all. Instead it interprets the difference between the days as a day and then displays that day's number; in this case, `19`.

To get the number of days until something happens, you have to do this:

```js
const expiry = moment( '2023-10-05 00:00Z' );
console.log( expiry.diff( moment(), 'days' ) );
```

Which outputs `50`, the number of days between now and 5 October.

Given that, I don't see how this has ever worked correctly or how it's working now. 

You'll notice that the other strings in this file, like [this one](https://github.com/Automattic/wp-calypso/blob/d17dd34d7c22a17991e0ad2d7bdca773a6d50b5d/client/me/purchases/manage-purchase/notices.tsx#L139), use `momentObject.fromNow()`. If I do this:

```js
const expiry = moment( '2023-10-05 00:00Z' );
console.log( expiry.fromNow() );
```

I get `in 2 months`, which is correct and looks like what you're aiming for here, but `daysToExpiry` does not look like it has the right value.

That said, I can't explain how you're getting the screenshots you've provided. Given the above information, what do you think is happening?

TheSteveK replied:
Just dropping by to confirm, I also see the correct text 

![Screenshot 2023-08-16 at 8 55 39 AM](https://github.com/Automattic/wp-calypso/assets/12895386/1893deac-579a-41c0-ba5a-695316e00ad0)

sirbrillig replied:
I see something similar, but in my case, the text is `in a year`, which is clearly not the string we're looking at here, which includes a hard-coded `days` word.

<img width="1074" alt="Screenshot 2023-08-16 at 10 40 23 AM" src="https://github.com/Automattic/wp-calypso/assets/2036909/cff1509a-37ef-4218-94f2-b06338924865">

I believe what we're both seeing is the text from [this line](https://github.com/Automattic/wp-calypso/blob/2b9d6aa97423edc312f9790014c9e6e02c2efdde/client/me/purchases/manage-purchase/notices.tsx#L127), which uses `expiry.fromNow()` as I mentioned above.

The reason is that this code is inside the `isMonthly()` check and the testing instructions don't mention that the product must be monthly to test these changes. If I modify the code to use this version, here's what I see:

<img width="630" alt="Screenshot 2023-08-16 at 10 44 53 AM" src="https://github.com/Automattic/wp-calypso/assets/2036909/7c9412d5-f4bc-400d-9ef0-c1be57c23704">

Note that even apart from the strange wording, the "1" is incorrect; this subscription expires in about 365 days.

This is definitely wrong and the one below it (in trunk) is also broken.

sirbrillig commented on client/me/purchases/manage-purchase/notices.tsx:
```diff
@@ -103,6 +103,15 @@ class PurchaseNotice extends Component<
+			}
+
 			return translate(
 				'%(purchaseName)s will expire and be removed from your site %(expiry)s days. ',
```
I know this is an existing string and not part of this diff, but it seems to suffer from the same issue I mentioned above. 

"Product will expire and be removed 5 days" does not make much sense.

sirbrillig reviewed (changes requested):
Two issues:

1. `moment( expiry.diff( moment() ) ).format( 'D' )` does not give the number of days until now that the author intended (it looks like this bug was added 7 years ago in https://github.com/Automattic/wp-calypso/pull/10380). 

2. Several strings are missing the word "in".

DavidRothstein commented:
Tried to get my head around this one -- @sirbrillig I think you're right about the `moment( expiry.diff( moment() ) ).format( 'D' )` problem, although in practice the bug doesn't seem to actually occur now, since it would only occur if the subscription expires more than a month away (which is not when this code block runs)?

With this pull request, though, the bug could occur, although you might need a monthly siteless purchase that expires more than a month from now -- and I'm not sure how easy that is to get in practice.

In any case, would the best fix for both the above and for the missing "in" word just be to remove the `if ( isMonthly() )` section **entirely** -- it doesn't seem to serve any purpose except to unnecessarily duplicate code and introduce the bug :smile:  In that case, this pull request would only need to add one block of code, not two, and the entire code would basically look like this:

```
if ( isTemporarySitePurchase( purchase ) ) {
	return translate( '%(purchaseName)s will expire and be removed %(expiry)s.', {
		args: {
			purchaseName: getName( purchase ),
			expiry: expiry.fromNow(),
		},
	} );
}

return translate( '%(purchaseName)s will expire and be removed from your site %(expiry)s.', {
	args: {
		purchaseName: getName( purchase ),
		expiry: expiry.fromNow(),
	},
} );
```

Everything would then be using `expiry.fromNow()`, which fixes the bug and also takes care of adding the word "in".

Commit:
Use isMonthlyPurchase instead of isMonthly (to correctly identify individual products)

Commit:
Correctly format the number of days to expiry

mattgawarecki commented:
@DavidRothstein You beat me to the punch by a few minutes 🙂 

What I found was this: `isMonthly` only works for _plans_, not _products_. So, to experience the bug, you have to have purchased a monthly plan like Security (10GB) or Complete.

I left the code block in place, but I replaced the call to `isMonthly` with `isMonthlyPurchase`, which accounts for both plans and individual products. I also added "in" to the appropriate spots.

Moving everything to `fromNow` could also work, but since that method uses "fuzzy time," it would lose some precision for folks on monthly billing terms. For example, my understanding of [the docs](https://momentjs.com/docs/#/displaying/fromnow/) is that anything 26+ days in the future would just show "in a month."

sirbrillig commented on client/me/purchases/manage-purchase/notices.tsx:
```diff
@@ -100,20 +100,41 @@ class PurchaseNotice extends Component<
+
+			if ( isTemporarySitePurchase( purchase ) ) {
+				return translate(
+					'%(purchaseName)s will expire and be removed in %(daysToExpiry)d days. ',
```
Back to my original minor comment: there's an extra space here.

```suggestion
					'%(purchaseName)s will expire and be removed in %(daysToExpiry)d days.',
```

sirbrillig commented on client/me/purchases/manage-purchase/notices.tsx:
```diff
@@ -100,20 +100,41 @@ class PurchaseNotice extends Component<
 
 			return translate(
-				'%(purchaseName)s will expire and be removed from your site %(expiry)s days. ',
+				'%(purchaseName)s will expire and be removed from your site in %(daysToExpiry)d days. ',
```
Same space here.

```suggestion
				'%(purchaseName)s will expire and be removed from your site in %(daysToExpiry)d days.',
```

sirbrillig reviewed (commented):
This looks much better. Thank you for the fix! Just a couple minor things but otherwise this looks great and fixes the bugs we didn't even know we had.

Commit:
Remove unnecessary trailing spaces from expiry text

sirbrillig reviewed (approved):
From a Payments perspective, this looks good to me!

spraveenitpro commented:
<details>
<summary>Without patch:</summary>

![eNTYb1.png](https://github.com/Automattic/wp-calypso/assets/552016/5d3b053a-afdb-4214-a0d1-296de08bb967)
</details>

<details>
<summary>With Patch:</summary>

![YU6D8y.png](https://github.com/Automattic/wp-calypso/assets/552016/ccaa7935-2cd0-4d0e-9e3f-fac30aaf503d)
</details>

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8799746

Thank you @mattgawarecki for including a screenshot in the description! This is really helpful for our translators.
